components/script/dom/htmlimageelement.rs

```diff
@@ -356,20 +390,253 @@ impl HTMLImageElement {
     }
 
     /// <https://html.spec.whatwg.org/multipage/#update-the-source-set>
-    fn update_source_set(&self) -> Vec<DOMString> {
+    fn update_source_set(&self) {
+        // Step 1
+        *self.source_set.borrow_mut() = HTMLImageElement::new_source_set();
+
+        // Step 2
         let elem = self.upcast::<Element>();
-        // TODO: follow the algorithm
-        let src = elem.get_string_attribute(&local_name!("src"));
-        if src.is_empty() {
-            return vec![]
+        let parent = elem.upcast::<Node>().GetParentElement();
+        let is_parent_picture = match parent.clone() {
+            Some(xparent) => xparent.is::<HTMLPictureElement>(),
+            None => false
+        };
+        let nodes = parent.unwrap().upcast::<Node>().ChildNodes();
+        let elements = match is_parent_picture {
+            true => nodes.iter().map(|n| DomRoot::from_ref(&*(*n).downcast::<Element>().unwrap())).collect(),
```

Comment:
Is there any guarantee that the child nodes are `Element`s? I suspect you can end up with text nodes there easily, and then this will crash.

Reply:
(So you should probably do `filter_map(DomRoot::downcast::<Element>)` instead).

Reply:
Sure will update 👍